### PR TITLE
MLIR test for ZNE local folding

### PR DESCRIPTION
**Pull request by cosenal**
A draft PR (⚠️ **don't merge**) with an MLIR test that should serve as a guide through the implementation of local folding in the ZNE transform of the Mitigation dialect. 

The IR added in this test should correspond to the lowering of the following Catalyst frontend code:
```python3
dev = qml.device("lightning.qubit", wires=2)

@qml.qnode(dev)
def circuit():
    qml.Hadamard(wires=0)
    qml.CNOT(wires=[0, 1])
    return qml.expval(qml.PauliY(wires=0))

@qjit(keep_intermediate=True)
def mitigated_circuit():
    s = jax.numpy.array([1, 2, 3])
    return mitigate_with_zne(circuit, scale_factors=s, folding='all')()
```

At the moment this test is not supposed to pass, because the logic implementation is not there for local folding yet.

**Related GitHub Issue:** https://github.com/PennyLaneAI/catalyst/issues/755

### CODE REVIEW

File: mlir/test/Mitigation/ZneFoldingAllTest.mlir
Comment (dime10) [inline mlir/test/Mitigation/ZneFoldingAllTest.mlir:1]: Looks like a good starting point! The way this suggests the pass would be implemented looks good to me.
The only thing I would suggest is to follow these best-practices for writing FileCheck tests: https://mlir.llvm.org/getting_started/TestingGuide/#filecheck-best-practices

Right now, the test is too specific as it checks almost verbatim what the output IR should look like.

Comment (cosenal) [reply mlir/test/Mitigation/ZneFoldingAllTest.mlir:1]: Thanks for the pointer, @dime10!
I incorporated those best practices in the local-folding test, and also took the chance to port some of them to the existing global-folding test. Please take another look.

File: mlir/test/Mitigation/ZneFoldingAllTest.mlir [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    //CHECK:    [[results:%.+]] = scf.for [[idx:%.+]] = [[c0]] to [[c3]] step [[c1]] iter_args(%arg1 = [[emptyRes]]) -> (tensor<3xf64>) {
+        //CHECK:    [[scaleFactor:%.+]] = tensor.extract [[dense3]][[[idx]]] : tensor<3xindex>
+        //CHECK:    [[intermediateRes:%.+]] = func.call @circuit.folded([[scaleFactor]]) : (index) -> tensor<f64>
+        //CHECK:    [[tensorRes:%.+]] = tensor.from_elements [[intermediateRes]] : tensor<1xf64>
+        //CHECK:    [[resultsFor:%.+]] = scf.for %arg2 = [[c0]] to [[c1]] step [[c1]] iter_args(%arg3 = %arg1) -> (tensor<3xf64>) {
+            //CHECK:    [[extracted:%.+]] = tensor.extract [[tensorRes]][%arg3] : tensor<1xf64>
+            //CHECK:    [[insertedRes:%.+]] = tensor.insert [[extracted]] into %arg3[%arg1] : tensor<5xf64>
+            //CHECK:    scf.yield [[insertedRes]]
+        //CHECK:    scf.yield [[resultsFor]]
+    //CHECK:    return [[results]]
```
Comment (dime10) [inline mlir/test/Mitigation/ZneFoldingAllTest.mlir:82]: Nit: Personally I think the following format is slightly more readable (indentation is definitely useful):
```suggestion
    //CHECK:    [[c0:%.+]] = index.constant 0
    //CHECK:    [[c1:%.+]] = index.constant 1
    //CHECK:    [[c3:%.+]] = index.constant 3
    //CHECK:    [[dense3:%.+]] = arith.constant dense<[1, 2, 3]>
    //CHECK:    [[emptyRes:%.+]] = tensor.empty() : tensor<3xf64>
    //CHECK:    [[results:%.+]] = scf.for [[idx:%.+]] = [[c0]] to [[c3]] step [[c1]] iter_args(%arg1 = [[emptyRes]]) -> (tensor<3xf64>) {
    //CHECK:        [[scaleFactor:%.+]] = tensor.extract [[dense3]][[[idx]]] : tensor<3xindex>
    //CHECK:        [[intermediateRes:%.+]] = func.call @circuit.folded([[scaleFactor]]) : (index) -> tensor<f64>
    //CHECK:        [[tensorRes:%.+]] = tensor.from_elements [[intermediateRes]] : tensor<1xf64>
    //CHECK:        [[resultsFor:%.+]] = scf.for %arg2 = [[c0]] to [[c1]] step [[c1]] iter_args(%arg3 = %arg1) -> (tensor<3xf64>) {
    //CHECK:            [[extracted:%.+]] = tensor.extract [[tensorRes]][%arg3] : tensor<1xf64>
    //CHECK:            [[insertedRes:%.+]] = tensor.insert [[extracted]] into %arg3[%arg1] : tensor<5xf64>
    //CHECK:            scf.yield [[insertedRes]]
    //CHECK:        scf.yield [[resultsFor]]
    //CHECK:    return [[results]]
```

**Review by dime10 — COMMENTED** [review]
Thank you for the updates, I have a few more suggestions from the best practices :)

**Comment by cosenal** [comment]
Closing in favor of https://github.com/PennyLaneAI/catalyst/pull/1006, which this branch has been merged into.
We will continue there, as it makes more sense to write code and tests in the same PR.

**Pull request closed without merging**